hazelcast/src/main/java/com/hazelcast/query/impl/predicates/MultiPartitionPredicateImpl.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.query.impl.predicates;
+
+import static com.hazelcast.internal.util.Preconditions.checkNotNull;
+import static com.hazelcast.internal.util.Preconditions.checkTrue;
+
+import com.hazelcast.internal.serialization.BinaryInterface;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.nio.serialization.impl.Versioned;
+import com.hazelcast.query.PartitionPredicate;
+import com.hazelcast.query.Predicate;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Random;
+import java.util.Set;
+
+/**
+ * Implementation of {@link PartitionPredicate}.
+ *
+ * @param <K> type of the entry key
+ * @param <V> type of the entry value
+ */
+@BinaryInterface
+public class MultiPartitionPredicateImpl<K, V> implements PartitionPredicate<K, V>, IdentifiedDataSerializable, Versioned {
```

Comment:
This shouldn't be `Versioned` anymore

Reply:
A potential issue here: 
What happens when i pass `MultiPartitionPredicateImpl`, for example, to `executeOnEntries` method in a cluster with version 5.1? I expect it not to work.

Reply:
Why should it work? I believe it is okay for it to not work in this case, as the feature is implemented in 5.2. Similarly, if you send this predicate to 5.0 member from 5.2 client, I believe it shouldn't work as well (which is also the case with the current implementation)

Reply:
I expect it not to work, my above comment is saying that.

Reply:
Yeah, I see. I am just saying that I don't see this as a potential issue

Reply:
Good point, I will remove this.

Reply:
@software-is-art What about this one?
> What happens when i pass MultiPartitionPredicateImpl, for example, to executeOnEntries method in a cluster with version 5.1?

Do we throw feature is not supported exception in this case?

Reply:
Is there a standard way of doing that? Would I just be checking for the destination cluster version before sending it?

Reply:
You can check `MultiMapProxyImpl#putAllAsync` method to see how this control can be done.
Basically you need to do this control for source cluster entry points where one can pass multi-partition predicate, including client-message tasks.

I think, to address this you can create a separate PR later, after this one is merged, since this one is already got bloated.

Reply:
That sounds good to me, happy to have that in a separate PR